pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -250,6 +256,21 @@ func (s *FlinkStateMachine) handleNewOrUpdating(ctx context.Context, application
 		// Reset teardown hash if set
 		application.Status.TeardownHash = ""
 	}
+
+	// Reset the in-place update hash if set
+	hash := flink.HashForApplication(application)
+	if application.Status.InPlaceUpdatedFrom == hash {
```

Comment:
Should we also check for `scaleMode` here and only apply it to the `InPlace` mode? The (possibly contrived) scenario I'm thinking of is when a user moves from 
- P -> P' using  the `InPlace` mode 
- Makes a subsequent update to `NewCluster` mode
- Moves back to P. 

I _think_ we'd still block that according to this condition right?

Reply:
Yeah, good catch. We should only block if scaleMode is InPlace. I've added that check.

Reply:
Why is that the case? Intuitively it seems like if the current scaleMode is NewCluster, `InPlaceUpdatedFrom` should become irrelevant right? And the deployment should be able to proceed as a regular update?

Reply:
I tried to explain the problem in the comment below, but it's a bit complicated. This is basically an edge case when you go from P -> P' -> P, where the P -> P' was in-place. That means that our cluster (e.g., our deployments) are are associated with Hash(P). Now we're trying to go _back_ to P (our original parallelism) which means that we need to create a new set of deployments for Hash(P). But we can't do that, because they already exist (and we're using them). We also can't scale down the existing ones (trying to "recreate" the original setup) because that would disrupt the original job.

The only way to fix the situation is to get the user to change something in the spec that will give us a new hash, allowing the deployment to proceed. Hopefully this is an obscure enough case that users won't hit it, but if they do we need to protect against it.

Reply:
Okay that makes sense! Thanks for explaining. I wonder if adding `scaleMode` to the application hash would be a way to automatically get us a new hash (at least for the above scenario).

Reply:
Yeah I don't think adding scaleMode into the hash is a good idea, as that would prevent you from enabling in-place and doing an in-place deploy without causing a full deploy to happen.